File: leetcode_python/75.py
"""
Given an array with n objects colored red, white or blue, sort them in-place so that objects of the same color are adjacent, with the colors in the order red, white and blue.

Here, we will use the integers 0, 1, and 2 to represent the color red, white, and blue respectively.

Note: You are not suppose to use the library's sort function for this problem.

Example:

Input: [2,0,2,1,1,0]
Output: [0,0,1,1,2,2]
"""


class Solution:
    def sortColors(self, nums):
        zero_index = 0
        two_index = len(nums) - 1
        index = 0
        while index <= two_index:
            if nums[index] == 0:
                nums[index], nums[zero_index] = nums[zero_index], nums[index]
                zero_index += 1
                index += 1
            elif nums[index] == 1:
                index += 1
            elif nums[index] == 2:
                nums[index], nums[two_index] = nums[two_index], nums[index]
                two_index -= 1


# nums = [2, 0, 2, 1, 1, 0]
nums = [1, 2, 0]
Solution().sortColors(nums)
print(nums)

File: leetcodeJS/75.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */

 var sortColors = function(nums) {
    var left = 0,
        right = nums.length - 1;

    while (nums[left] == 0) {
        left++;
    }

    while (nums[right] == 2) {
        right--;
    }

    var index = left;

    while (index <= right) {
        if (nums[index] == 0) {
            [nums[left], nums[index]] = [nums[index], nums[left]];
            left += 1;
        }
        if (nums[index] == 2) {
            [nums[index], nums[right]] = [nums[right], nums[index]];
            right -= 1;
        } else {
            index += 1;
        }
    }
};

 var sortColorsStupid = function(nums) {
    var red = 0,
        white = 0,
        blue = 0;
    for (var i = 0; i < nums.length; i++) {
        if (nums[i] == 0) {
            red += 1;
        } else if (nums[i] == 1) {
            white += 1;
        } else {
            blue += 1;
        }
    }

    var index = 0;
    while (index < red) {
        nums[index] = 0;
        index += 1;
    }
    while (index < red + white) {
        nums[index] = 1;
        index += 1;
    }
    while (index < nums.length) {
        nums[index] = 2;
        index += 1;
    }

};
